### PR TITLE
Add inputs tracking

### PR DESCRIPTION
**Pull request by Kuraiou**
1. Adds class-variable caching of the list of required inputs/and outputs.
2. updates `input` to return the list of required inputs when called without arguments, and aliases it to `inputs`
3. Adds `optional_inputs` method which returns the list of optional inputs.

### CODE REVIEW

File: lib/verbalize/version.rb [hunk 1]
```diff
@@ -1,3 +1,3 @@
 module Verbalize
-  VERSION = '2.0.1'.freeze
+  VERSION = '2.0.2'.freeze
```
Comment (tmertens) [inline lib/verbalize/version.rb]: 1. Version updates should be done immediately prior to a release, it makes it easier to track what is included in the release itself if it is the last commit.
2. This should be a minor version update, not a patch level update as it adds functionality and is not a bugfix.

File: lib/verbalize/action.rb [hunk 2]
```diff
@@ -17,8 +17,19 @@ def self.included(target)
...
       def input(*required_keywords, optional: [])
+        # treat `input` as an accessor if nothing is passed in.
+        return @inputs || [] if required_keywords.empty? && optional.empty?
+
+        @inputs = required_keywords
+        @optional_inputs = optional
+
         class_eval Build.call(required_keywords, Array(optional))
       end
+      alias_method :inputs, :input
```
Comment (tmertens) [inline lib/verbalize/action.rb]: Please update the README to document the new functionality.

TODO: @taylorzr We should consider adding auto-generated yard or relish documentation to the project

File: lib/verbalize/action.rb [hunk 3]
```diff
@@ -17,8 +17,19 @@ def self.included(target)
 
     module ClassMethods
       def input(*required_keywords, optional: [])
+        # treat `input` as an accessor if nothing is passed in.
+        return @inputs || [] if required_keywords.empty? && optional.empty?
```
Comment (taylorzr) [inline lib/verbalize/action.rb]: I would prefer to have the input introspection be only defined as **.inputs**, and not mix together defining inputs and listing inputs.

File: lib/verbalize/version.rb (hunk 1 above)
Comment (Kuraiou) [inline lib/verbalize/version.rb]: 👍   good point

File: lib/verbalize/action.rb (hunk 2 above)
Comment (Kuraiou) [inline lib/verbalize/action.rb]: oh yeah duh

File: lib/verbalize/action.rb (hunk 3 above)
Comment (Kuraiou) [inline lib/verbalize/action.rb]: sure.

File: spec/integration/action_spec.rb [hunk 4]
```diff
@@ -190,26 +190,31 @@ def call
...
+      it 'returns an empty array' do
+        expect(some_class.inputs).to eq []
+      end
     end
 
-    context 'with arguments' do
-      it 'allows arguments to be defined and delegates the class method \
-      to the instance method' do
-        some_class = Class.new do
+    context 'with required argumentsarguments' do
```
Comment (tmertens) [inline spec/integration/action_spec.rb]: `argumentsarguments`

File: lib/verbalize/action.rb (hunk 3 above)
Comment (tmertens) [inline lib/verbalize/action.rb]: I wonder if we should just make a separate `required_arguments` accessor method.  It's a bit unintuitive that `inputs` takes both required and optional inputs, but only returns one of the two.  I would prefer if `inputs` returned everything instead of separating them out, or if we used a `required_arguments` accessor method.  Or, we can do all three - `inputs` returns all of them as a hash, `(required|optional)_inputs` returns them separately.

File: spec/integration/action_spec.rb [hunk 5]
```diff
@@ -188,28 +188,33 @@ def call
     end
   end
 
-  describe '.input' do
+  describe '.input/.inputs' do
```
Comment (taylorzr) [inline spec/integration/action_spec.rb]: These should be separated :)

File: lib/verbalize/action.rb [hunk 6]
```diff
@@ -17,9 +17,20 @@ def self.included(target)
 
     module ClassMethods
       def input(*required_keywords, optional: [])
+        @inputs = required_keywords
+        @optional_inputs = optional
+
         class_eval Build.call(required_keywords, Array(optional))
       end
 
+      def inputs
```
Comment (tmertens) [inline lib/verbalize/action.rb]: My earlier comment got hidden by other changes:

I think we should name this `required_inputs` instead of `inputs`. It's a bit unintuitive that `input` takes both required and optional inputs, but `inputs` only returns one of the two.  We should use consistent naming and return all inputs here, or use appropriate method names if we are going to separate required/optional parameters.

Or, do all three - `inputs` returns required & optional parameters as a hash, while `(required|optional)_inputs` returns them individually.

Comment (tmertens) [inline lib/verbalize/action.rb]: Also @taylorzr I don't agree here with separating `input` and `inputs` at least for the purpose of accessing the full parameter list.

There is no contextual indicator of the difference between these two methods.  Conceptually, the original implementation made more sense, if we ignore that it returned less than the full set of inputs when used as a getter.

While it is generally good to have methods with individual responsibilities since it simplifies the logic of the accessor method itself, it is a common pattern to use the same method as a setter when params are given and getter when there are none.  It eliminates the ambiguity that currently exists between `input` and `inputs` which do not conform to either the setter-as-accessor pattern nor the typical `getter` / `setter=` accessor pattern we are used to.

TL;DR - if we are going to have a getter method for accessing the full input list directly, the original implementation here seems best (modified to return required+optional parameters).

File: README.md [hunk 7]
```diff
@@ -69,6 +69,32 @@ result = Divide.call(a: 1, b: 0) # => [:error, 'You can’t divide by 0']
...
+  def c
+    @c ||= 0
+  end
+
+  def d
+    @d ||= 0
+  end
+end
+
+Add.inputs # [:a, :b]
```
Comment (taylorzr) [inline README.md]: Documentation needs updated now :)

File: lib/verbalize/action.rb [hunk 8]
```diff
@@ -17,20 +17,24 @@ def self.included(target)
...
+      def required_inputs
+        @required_inputs || []
       end
 
       def optional_inputs
         @optional_inputs || []
       end
 
+      def inputs
+        required_inputs + optional_inputs
```
Comment (tmertens) [inline lib/verbalize/action.rb]: This mashes everything together into a single array.  I'm not sure that is what we really want here.  Maybe a hash with keys for each?  Or an array with the same format as the argument list to the `input` method?

It's hard to know what the right implementation is here since we don't have a concrete use case

Comment (taylorzr) [inline lib/verbalize/action.rb]: I think this is fine the way it is. 

I do agree it would be better if we had a clear use case, but I think this is so simple it's not going to be a problem.

File: README.md (hunk 7 above)
Comment (Kuraiou) [inline README.md]: GDI.

File: lib/verbalize/action.rb (hunk 8 above)
Comment (Kuraiou) [inline lib/verbalize/action.rb]: we have two use-cases in master, both of which want both required and optional inputs to slice from a larger hash.

**Comment by Kuraiou** [comment]
@taylorzr you're the one that has to merge, as I don't have write access into verbalize.

**Comment by taylorzr** [comment]
JFYI, Circle was previously not configured to run pull requests from forks. I changed Circle settings to enable this.
